src/CauserResolver.php

```diff
@@ -0,0 +1,115 @@
+<?php
+namespace Spatie\Activitylog;
+
+use Closure;
+use Illuminate\Auth\AuthManager;
+use Illuminate\Contracts\Config\Repository;
+use Illuminate\Database\Eloquent\Model;
+use InvalidArgumentException;
+use Spatie\Activitylog\Exceptions\CouldNotLogActivity;
+
+class CauserResolver
+{
+    protected AuthManager $authManager;
+
+    protected string $authDriver;
+
+    /**
+     * User defined model or callback to override default reslover logic
+     */
+    protected Model | Closure | null $override = null;
+
+    public function __construct()
+    {
+        $config = app(Repository::class);
+        $this->authManager = app(AuthManager::class);
+
+        $this->authDriver = $config['activitylog']['default_auth_driver'] ?? $this->authManager->getDefaultDriver();
+    }
+
+    /**
+     * Reslove causer based different arguments
+
+     * @param Model|int|Closure|null $subject
+     * @return null|Model
+     * @throws InvalidArgumentException
+     * @throws CouldNotLogActivity
+     */
+    public function resolve(Model | int | Closure | null $subject = null) : ?Model
+    {
+        if ($subject instanceof Closure) {
+            $this->override = $subject;
+        }
+
+        if (! is_null($this->override) || is_null($subject)) {
+            return $this->getCauser();
+        }
+
+        if ($this->isValidCauser($subject)) {
+            return $subject;
+        }
+
+
+        // Resolve the user based on passed id
+        $guard = $this->authManager->guard($this->authDriver);
+
+        $provider = method_exists($guard, 'getProvider') ? $guard->getProvider() : null;
+        $model = method_exists($provider, 'retrieveById') ? $provider->retrieveById($subject) : null;
+
+        if ($model instanceof Model) {
+            return $model;
+        }
+
+        throw CouldNotLogActivity::couldNotDetermineUser($subject);
+    }
+
+    protected function getCauser(): ?Model
+    {
+        if (! $this->isValidOverrideType()) {
+            return $this->getDefaultCauser();
+        }
+
+        $causer = $this->override;
+
+        if ($this->override instanceof Closure) {
+            $causer = ($this->override)();
+        }
+
+        if (! $this->isValidCauser($causer)) {
+            throw CouldNotLogActivity::couldNotDetermineUser($causer);
+        }
+
+        return $causer;
+    }
+
+    public function withResolver(Closure $callback): self
+    {
+        $this->override = $callback;
```

Comment:
The resolver closure should be separated from the enforced causer instance.
It can be that you want to define both and in that case you have a race-condition. This shouldn't be the case but both should be saved and the enforced one will be returned.

```suggestion
        $this->resolver = $callback;
```

Reply:
- In what use case/scenario using both inforced causer instance and callback would be applicable?
- Who should win the race-condition?

I see current implementation is reasonable
- enforcing the callback when ever available.
- use `setCauser()` method when ever we want to use a causer instance
- use `resolveUsing()` method whenever we want to use callback to resolve the causer whether it's a model or null.

In this case we'll not race-condition at all. I see using both doesn't make since in this context.